### ISSUE TITLE
Hitting catalog_product_entity_varchar maximum auto increment ID for complex catalog

### ISSUE THREAD

**Issue by convenient** [magento/magento2#21890]
## Preconditions
1. Magento 2.1 (this issue is from EE but the bug is in CE and will be affecting other parts of the system)
2. Multi store website.
3. Product has global attributes assigned.

## Expected result

Triggering the same product save multiple times should not affect the data or structure within `catalog_product_entity_varchar`, the next auto increment `value_id` should be +1 of the current `max(value_id)`

## Actual result

The auto increment ticks up each save request by the number of global attributes assigned to the product, wasting valid numbers. The data inside the table does not change.

At scale this wastage of increment ids can result in the table maxing out and breaking all product updates.

## Steps to reproduce and more details

The below was discovered on a client running 2.1 EE. I have verified that the issue has been fixed in 2.2 and 2.3 by [app/code/Magento/Catalog/Plugin/Model/ResourceModel/ReadSnapshotPlugin.php](https://github.com/magento/magento2/blob/2.3.0/app/code/Magento/Catalog/Plugin/Model/ResourceModel/ReadSnapshotPlugin.php)

Due to the way the staging module works with the versioned products a number of auto increment ids may be wasted each time a product is saved in a store scope.

All the below focuses on the `catalog_product_entity_varchar` table but it should apply to each of the backend types, provided there are enough attributes and store views defined.

I know EE issues may not be considered relevant for this board but I'm posting it here for the following reasons
- The core issue in itself is that 2.1 does not have a `ReadSnapshotPlugin` which means the snapshot being loaded will contain missing data for global attributes. 
- This is such an unusual problem that posting it here ensures it will be picked up by Google and it may help other developers searching it in the future. Even if this issue is closed it will provide value for other developers hopefully.

### To replicate

In a fresh install of 2.1EE

1. Log into the admin panel.

1. Go `Stores -> All Stores` and create a new website,  store, and store view so that it looks like this.

    <img width="1319" alt="Screenshot 2019-03-21 at 08 32 52" src="https://user-images.githubusercontent.com/600190/54740677-fdc0a200-4bb3-11e9-83a2-a2f2db21d9da.png">

1. Go `Stores -> Attributes -> Products` find the `name` attribute, under `Advanced Attribute Properties` make it `Scope=Global`, save the attribute.

1. Go `System -> Extensions -> Integrations`, add a new integration, give it all resources, save it, activate it, and copy the `Access Token`.

1. Run the following `curl` command to create a simple product (after replacing the access token with your access token).
    
    ```bash
    curl -s -k -X PUT https://magento-ee-21.localhost/rest/all/V1/products/productsku -H 'authorization: Bearer YOUR_TOKEN_HERE' -H 'cache-control: no-cache' -H 'content-type: application/json' -d '{"product":{"sku":"productsku","name":"product name","attribute_set_id":4,"type_id":"simple","price":100,"status":1}}'
    ```

1. Run the following `curl` command to set the store scoped `uk` values (after replacing the access token with your access token).
    
    ```bash
    curl -s -k -X PUT https://magento-ee-21.localhost/rest/uk/V1/products/productsku -H 'authorization: Bearer YOUR_TOKEN_HERE' -H 'cache-control: no-cache' -H 'content-type: application/json' -d '{"product":{"sku":"productsku","name":"product name","attribute_set_id":4,"type_id":"simple","price":100,"status":1}}'
    ```
 
1. Run the following queries directly in the database

    ```sql
     # See that all varchar data is duplicated in store scope except for "product name" (attribute 73) which is global.
     mysql> select * from catalog_product_entity_varchar;
     +----------+--------------+----------+--------+--------------+
     | value_id | attribute_id | store_id | row_id | value        |
     +----------+--------------+----------+--------+--------------+
     |        1 |           73 |        0 |      1 | product name |
     |        2 |          106 |        0 |      1 | container2   |
     |        3 |          124 |        0 |      1 | product-name |
     |        4 |          174 |        0 |      1 | 2            |
     |        6 |          106 |        2 |      1 | container2   |
     |        7 |          124 |        2 |      1 | product-name |
     |        8 |          174 |        2 |      1 | 2            |
     +----------+--------------+----------+--------+--------------+
     7 rows in set (0.00 sec)   
    
    
    # Get the next auto increment id, see that its +1 of the max value
    mysql> SELECT AUTO_INCREMENT FROM INFORMATION_SCHEMA.TABLES WHERE table_name = 'catalog_product_entity_varchar' and table_schema='ee2.1';
    +----------------+
    | AUTO_INCREMENT |
    +----------------+
    |              9 |
    +----------------+
    1 row in set (0.00 sec)
   ```

1. Re-run the previous store scoped `curl` command, as all values are already complete we should expect that the current `value_id` data remain unchanged.
   
    ```bash
    curl -s -k -X PUT https://magento-ee-21.localhost/rest/uk/V1/products/productsku -H 'authorization: Bearer YOUR_TOKEN_HERE' -H 'cache-control: no-cache' -H 'content-type: application/json' -d '{"product":{"sku":"productsku","name":"product name","attribute_set_id":4,"type_id":"simple","price":100,"status":1}}'
    ```

1. Re-run the queries you did above and compare the results, see that the auto increment ID is increasing by 1 each time.

    ```sql
     # See that all the varchar data is unchanged
     mysql> select * from catalog_product_entity_varchar;
     +----------+--------------+----------+--------+--------------+
     | value_id | attribute_id | store_id | row_id | value        |
     +----------+--------------+----------+--------+--------------+
     |        1 |           73 |        0 |      1 | product name |
     |        2 |          106 |        0 |      1 | container2   |
     |        3 |          124 |        0 |      1 | product-name |
     |        4 |          174 |        0 |      1 | 2            |
     |        6 |          106 |        2 |      1 | container2   |
     |        7 |          124 |        2 |      1 | product-name |
     |        8 |          174 |        2 |      1 | 2            |
     +----------+--------------+----------+--------+--------------+
     7 rows in set (0.00 sec)   
    
    
    # See that the auto increment ID has ticked up by +1 erroneously.
    mysql> SELECT AUTO_INCREMENT FROM INFORMATION_SCHEMA.TABLES WHERE table_name = 'catalog_product_entity_varchar' and table_schema='ee2.1';
    +----------------+
    | AUTO_INCREMENT |
    +----------------+
    |             10 |
    +----------------+
    1 row in set (0.00 sec)
   ```

1. Every time you run the `uk` store view `curl` command you will see the auto increment ID jump ahead but the actual data in the `catalog_product_entity_varchar` table remaining unchanged. For this example with a single attribute it has little effect, but at scale this is a bigger problem.

### Real Life Data

When a product save is issued for a store view the eav backend tables waste a number of auto increment `value_id`s proportional to the number of global attributes assigned to the product in the set.

With large catalogs this means we are wasting a large proportion of the potential auto increment ids. When the increment IDs max out `2147483648` product updates begin failing.

Some details of the size of the catalog.

| Detail        | Value           | SQL  |
| ------------- |:-------------:| -----:|
| Numbers of SKUs      | ~400,000 | `select count(distinct sku) from catalog_product_entity;` |
| Average number of global `varchar` attributes per set | 60 | `select avg(number_of_attributes_per_set) from (select count(attribute_id) as number_of_attributes_per_set from eav_entity_attribute where attribute_id in (select attribute_id from eav_attribute where entity_type_id=4 and backend_type='varchar') and attribute_id in(select attribute_id from catalog_eav_attribute where is_global=1) group by attribute_set_id) as T;`|
| Number of store views | 12 | `select count(*)-1 from store;`

As there are 60 global varchar attributes per product there will be 60 auto increment IDs wasted each time this product is saved in a store scope.

Approximately once a month a price update goes through for a lot of the catalog, price is stored on a store view level.

As each save on each store will trigger this wastage of increment IDs we can work out how many full price pushes we have until we run out of the available mysql `int(11)` max value.

`(2^31)/(12 * 400000 * 60)=7.45`

Thats just over 7 full pushes to the entire catalog to break the varchar tables, thus breaking product updates. (It's not quite as bad as this in real life as it's not a full push every month, but this is a worst case calculation).

We had to do some unexpected workarounds to get the database back in order, see the bottom of this issue for our workaround.

### What is happening in the code

1. `Magento\Catalog\Model\ProductRepository::save`
    - Loads the product with all associated store scope data (falling back to scope 0 where relevant)
    - Triggers a save of the product
1. `Magento\Staging\Model\Operation\Update::processUpdate`
    - is triggered as an after commit callback
1. `Magento\Staging\Model\Operation\Update\UpdateEntityVersion::execute`
    - Is responsible for updating the main entity, attributes, and extensions.
1. `Magento\Eav\Model\ResourceModel\UpdateHandler::execute`
    - The attribute updater. It loads a snapshot of the current eav values from the database for the current store view. 
    - ⚠️ As there is no `ReadSnapshotPlugin` defined in 2.1 this will contain no global attributes. ⚠️ 
    - As this entity has global values set against it when the product is loaded but not against the current snapshot it will register these as inserts
1. `Magento\Eav\Model\ResourceModel\AttributePersistor::processInserts`
    - This function will process an insert for each global attribute
    - The query will be like `INSERT INTO 'catalog_product_entity_varchar' ('row_id','attribute_id','value','store_id') VALUES (?, ?, ?, ?) ON DUPLICATE KEY UPDATE 'row_id' = VALUES('row_id'), 'attribute_id' = VALUES('attribute_id'), 'value' = VALUES('value'), 'store_id' = VALUES('store_id')`.
    - As the key will be duplicate (because its a global attribute) it will update the existing row with the same data, and mysql will count +1 on the auto increment value

### Workaround

- We were only utilising about 1.5% of the available rows, but had burned through nearly all the increment IDs. The following actions rectified the issue for now.

- `show create table catalog_product_entity_varchar` to get the create table statement. 
- Modify the name of the table and the constraints to have `_tmp` appended.

```sql
CREATE TABLE `catalog_product_entity_varchar_tmp` (
  `value_id` int(11) NOT NULL AUTO_INCREMENT COMMENT 'Value ID',
  `attribute_id` smallint(5) unsigned NOT NULL DEFAULT '0' COMMENT 'Attribute ID',
  `store_id` smallint(5) unsigned NOT NULL DEFAULT '0' COMMENT 'Store ID',
  `row_id` int(10) unsigned NOT NULL COMMENT 'Version Id',
  `value` varchar(255) DEFAULT NULL COMMENT 'Value',
  PRIMARY KEY (`value_id`),
  UNIQUE KEY `CATALOG_PRODUCT_ENTITY_VARCHAR_ENTITY_ID_ATTRIBUTE_ID_STORE_ID` (`row_id`,`attribute_id`,`store_id`),
  KEY `CATALOG_PRODUCT_ENTITY_VARCHAR_ATTRIBUTE_ID` (`attribute_id`),
  KEY `CATALOG_PRODUCT_ENTITY_VARCHAR_STORE_ID` (`store_id`),
  CONSTRAINT `CATALOG_PRODUCT_ENTITY_VARCHAR_STORE_ID_STORE_STORE_ID_TMP` FOREIGN KEY (`store_id`) REFERENCES `store` (`store_id`) ON DELETE CASCADE,
  CONSTRAINT `CAT_PRD_ENTT_VCHR_ATTR_ID_EAV_ATTR_ATTR_ID_TMP` FOREIGN KEY (`attribute_id`) REFERENCES `eav_attribute` (`attribute_id`) ON DELETE CASCADE,
  CONSTRAINT `CAT_PRD_ENTT_VCHR_ROW_ID_CAT_PRD_ENTT_ROW_ID_TMP` FOREIGN KEY (`row_id`) REFERENCES `catalog_product_entity` (`row_id`) ON DELETE CASCADE
) ENGINE=InnoDB AUTO_INCREMENT=19 DEFAULT CHARSET=utf8 COMMENT='Catalog Product Varchar Attribute Backend Table'
```

- Bring down the product API so that no product updates were occurring. 
- Create this new table and populated it with the following 

```sql
insert into catalog_product_entity_varchar_tmp(attribute_id,store_id,row_id,value) select attribute_id,store_id,row_id,value from catalog_product_entity_varchar;
```

- Swap the table names so that this new table was the main table for `catalog_product_entity_varchar`.
- Then rename the constraints back to their vanilla value (without `_TMP` at the end).
- We have set up an alert for when tables start approaching this max value so that we can deal with it manually, until we upgrade to 2.3.

```sql
show table status where Name!='staging_update' and Auto_increment > floor(power(2,31)*0.8)
```

**Comment by magento-engcom-team** [comment]
Hi @convenient. Thank you for your report. 
To help us process this issue please make sure that you provided the following information:
- [x] Summary of the issue
- [x] Information on your environment
- [x] Steps to reproduce
- [x] Expected and actual results

Please make sure that the issue is reproducible on the vanilla Magento instance following **Steps to reproduce**. To deploy vanilla Magento instance on our environment, please, add a comment to the issue: 

`@magento-engcom-team give me 2.3-develop instance` - upcoming 2.3.x release

For more details, please, review the [Magento Contributor Assistant](https://devdocs.magento.com/guides/v2.2/contributor-guide/contributing.html#contributor-assist) documentation.

@convenient do you confirm that you was able to reproduce the issue on vanilla Magento instance following steps to reproduce?
- [x] yes
- [ ] no

**Comment by magento-engcom-team** [comment]
Hi @engcom-backlog-nazar. Thank you for working on this issue. 
In order to make sure that issue has enough information and ready for development, please read and check the following instruction: :point_down:
- [ ] 1. Verify that issue has all the required information. (**Preconditions**, **Steps to reproduce**, **Expected result**, **Actual result**).<details><summary>Details</summary>If the issue has a valid description, the label `Issue: Format is valid` will be added to the issue automatically. Please, edit issue description if needed, until label `Issue: Format is valid` appears.</details>

- [ ] 2. Verify that issue has a meaningful description and provides enough information to reproduce the issue. If the report is valid, add `Issue: Clear Description` label to the issue by yourself.

- [ ] 3. Add `Component: XXXXX` label(s) to the ticket, indicating the components it may be related to.
- [ ] 4. Verify that the issue is reproducible on `2.3-develop` branch<details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.3-develop instance` to deploy test instance on Magento infrastructure. <br/> - If the issue is reproducible on `2.3-develop` branch, please, add the label `Reproduced on 2.3.x`.<br/>- If the issue is not reproducible, add your comment that issue is not reproducible and close the issue and **_stop verification process here_**!
</details>

- [ ] 5. Verify that the issue is reproducible on `2.2-develop` branch. <details><summary>Details</summary>- Add the comment `@magento-engcom-team give me 2.2-develop instance` to deploy test instance on Magento infrastructure. <br/>- If the issue is reproducible on `2.2-develop` branch, please add the label `Reproduced on 2.2.x`</details>

-  _Next steps are available in case you are a member of **Community Maintainers**._

- [ ] 6. Add label `Issue: Confirmed` once verification is complete. 

- [ ] 7. Make sure that automatic system confirms that report has been added to the backlog.

**Comment by ghost** [comment]
Hi @convenient thank you for you report, We no longer accept pull requests to the v2.1 or v2.0 release lines to focus all development efforts on v2.3 and v2.2. 
https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html  
Thanks for collaboration, This issue will help people who discovered this issue on 2.1 version

**Comment by convenient** [comment]
Thanks @engcom-backlog-nazar that's pretty much what I expected 👍

**Comment by pmonosolo** [comment]
@convenient has this issue been solved after 2.3? I'm here on 2.4.3 and just hit this problem.
